### PR TITLE
routerrpc+lncli: enable subserver by default and switch over lncli sendpayment

### PR DESCRIPTION
**Pull request by joostjager**
This PR removes the conditional compilation logic from `routerrpc`. It will be enabled by default. This is not a big change, as it was already enabled in our pre-built binary releases.

With `routerrpc` being available always, `lncli` can be switched over to using the `SendPayment` call on the sub-server.

The old main rpc `SendPayment` and `SendPaymentSync` calls are marked as deprecated and expected to be removed in the next major release 0.11.

Note: SendToRoute will be switched over too, but after #3970 lands.

### CODE REVIEW

File: subrpcserver_config.go [hunk 1]
```diff
@@ -219,21 +219,9 @@ func (s *subRPCServerConfigs) PopulateDependencies(cc *chainControl,
 				reflect.ValueOf(genInvoiceFeatures),
 			)
 
+		// RouterRPC isn't conditionally compiled and doesn't need to be
```
Comment (cfromknecht) [inline subrpcserver_config.go]: can't this case be removed altogether?

File: cmd/lncli/cmd_pay.go [hunk 2]
```diff
@@ -0,0 +1,581 @@
...
+
+// paymentFlags returns common flags for sendpayment and payinvoice.
+func paymentFlags() []cli.Flag {
+	return []cli.Flag{
+		cli.StringFlag{
+			Name:  "pay_req",
+			Usage: "a zpay32 encoded payment request to fulfill",
+		},
+		cli.Int64Flag{
+			Name: "fee_limit",
```
Comment (cfromknecht) [inline cmd/lncli/cmd_pay.go]: not blocking, but could be a good time to rename this to `fee_limit_sat` to match `fee_limit_percent` so they both have units

File: cmd/lncli/cmd_pay.go [hunk 3]
```diff
@@ -127,34 +133,31 @@ var sendPaymentCommand = cli.Command{
...
-		}, nil
+		return ctx.Int64("fee_limit"), nil
+
 	case ctx.IsSet("fee_limit_percent"):
-		return &lnrpc.FeeLimit{
-			Limit: &lnrpc.FeeLimit_Percent{
-				Percent: ctx.Int64("fee_limit_percent"),
-			},
-		}, nil
+		return amt * ctx.Int64("fee_limit_percent") / 100, nil
```
Comment (cfromknecht) [inline cmd/lncli/cmd_pay.go]: no strong preference really, but do we want to round up here, e.g. `(amt * fee_limit_percent + 99)/100`? could help with very small payments

File: cmd/lncli/commands.go [hunk 4]
```diff
@@ -2643,7 +2643,7 @@ func queryRoutes(ctx *cli.Context) error {
 		return fmt.Errorf("amt argument missing")
 	}
 
-	feeLimit, err := retrieveFeeLimit(ctx)
+	feeLimit, err := retrieveFeeLimitLegacy(ctx)
```
Comment (cfromknecht) [inline cmd/lncli/commands.go]: do we keep this around bc there is no send to route on routerrpc?

File: lnrpc/routerrpc/router_server.go [hunk 5]
```diff
@@ -485,35 +485,6 @@ func (s *Server) trackPayment(paymentHash lntypes.Hash,
 			status.State = state
 		}
 
-		// Extract the last route from the given list of HTLCs. This
-		// will populate the legacy route field for backwards
-		// compatibility.
-		//
-		// NOTE: For now there will be at most one HTLC, this code
```
Comment (cfromknecht) [inline lnrpc/routerrpc/router_server.go]: would imagine this should be part of a different pr, no?

File: subrpcserver_config.go (hunk 1 above)
Comment (joostjager) [reply subrpcserver_config.go]: It will error in the default clause

File: cmd/lncli/commands.go (hunk 4 above)
Comment (joostjager) [reply cmd/lncli/commands.go]: It is around for `QueryRoutes` for which there is no version on `routerrpc` yet.

File: lnrpc/routerrpc/router_server.go (hunk 5 above)
Comment (joostjager) [reply lnrpc/routerrpc/router_server.go]: This PR is the moment we make `routerrpc` official. I don't think we should make a field official that is non-sensical.

File: subrpcserver_config.go (hunk 1 above)
Comment (cfromknecht) [reply subrpcserver_config.go]: okay, i missed the default case before. sgtm

**Comment by joostjager** [comment]
Todo: switch over integration tests to `routerrpc.SendPayment` :stuck_out_tongue_closed_eyes: Investigating how much effort that is.

**Comment by joostjager** [comment]
Itests switched over. Revealed one bug where routerrpc wasn't setting the payment request, see `routerrpc: fix empty payment request bug`

File: lnrpc/routerrpc/router_server.go (hunk 5 above)
Comment (cfromknecht) [reply lnrpc/routerrpc/router_server.go]: okay, i buy that

File: cmd/lncli/commands.go (hunk 4 above)
Comment (cfromknecht) [reply cmd/lncli/commands.go]: maybe something to consider for 0.11, people still find the plural name confusing anyway

File: lnrpc/routerrpc/router_backend.go [hunk 6]
```diff
@@ -650,6 +650,7 @@ func (r *RouterBackend) extractIntentFromSendRequest(
 		)
 		payIntent.DestFeatures = payReq.Features
 		payIntent.PaymentAddr = payReq.PaymentAddr
+		payIntent.PaymentRequest = []byte(rpcPayReq.PaymentRequest)
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply lnrpc/routerrpc/router_backend.go]: scratch that, we use `len(payreq) == 0` on the invoice side of things, but this is on the sender side

[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/routerrpc/router_backend.go]: You got me there :)

File: lnrpc/rpc.proto [hunk 7]
```diff
@@ -503,24 +503,29 @@ service Lightning {
...
+    by clients of the REST proxy. Additionally, this RPC expects the
+    destination's public key and the payment hash (if any) to be encoded as
+    hex strings.
     */
     rpc SendPaymentSync (SendRequest) returns (SendResponse) {
         option (google.api.http) = {
             post: "/v1/channels/transactions"
             body: "*"
         };
+        option deprecated = true;
```
[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/rpc.proto]: I think so. We only offer one way of sending a payment.

[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/rpc.proto]: The prior async version (async as being able to stream multiple payments with one call), was a very premature optimization imo. I haven't even been able to reproduce performance benefits.

[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/rpc.proto]: We could add a convenience rpc on `routerrpc` that blocks until there is a final outcome.

File: lntest/itest/lnd_test.go [hunk 8]
```diff
@@ -1468,15 +1485,18 @@ func testPaymentFollowingChannelOpen(net *lntest.NetworkHarness, t *harnessTest)
 	// executed.
 	ctxt, cancel = context.WithTimeout(ctxb, defaultTimeout)
 	defer cancel()
-	_, err = net.Alice.SendPaymentSync(
+	stream, err := net.Alice.RouterClient.SendPayment(
```
[the remark this answers is not in the document]
Comment (joostjager) [reply lntest/itest/lnd_test.go]: nice, done

File: lntest/itest/lnd_test.go [hunk 9]
```diff
@@ -10505,90 +10534,51 @@ func testBidirectionalAsyncPayments(net *lntest.NetworkHarness, t *harnessTest)
...
+	statusChan := make(chan *routerrpc.PaymentStatus)
 	for i := 0; i < numInvoices; i++ {
-		aliceSendReq := &lnrpc.SendRequest{
-			PaymentRequest: bobPayReqs[i],
-		}
-
-		bobSendReq := &lnrpc.SendRequest{
-			PaymentRequest: alicePayReqs[i],
-		}
+		send := func(node *lntest.HarnessNode, payReq string) {
```
[the remark this answers is not in the document]
Comment (joostjager) [reply lntest/itest/lnd_test.go]: moved

File: lnrpc/rpc.proto (hunk 7 above)
[the remark this answers is not in the document]
Comment (joostjager) [reply lnrpc/rpc.proto]: Offline discussion: convenience method isn't enough, because also rest wrappers are needed on `routerrpc`. For now decided to only deprecated the async send payment call in this PR.

File: cmd/lncli/cmd_pay.go [hunk 10]
```diff
@@ -303,7 +302,8 @@ func sendPaymentRequest(ctx *cli.Context, req *lnrpc.SendRequest) error {
 		req.LastHopPubkey = lastHop[:]
 	}
 
-	req.CltvLimit = uint32(ctx.Int(cltvLimitFlag.Name))
+	req.CltvLimit = int32(ctx.Int(cltvLimitFlag.Name))
```
Comment (bhandras) [inline cmd/lncli/cmd_pay.go]: Is there a reason not using `ctx.Uint32` here? (as `CltvLimit` is `uint32`)

File: cmd/lncli/cmd_pay.go [hunk 11]
```diff
@@ -334,54 +334,73 @@ func sendPaymentRequest(ctx *cli.Context, req *lnrpc.SendRequest) error {
...
-	// If we get a payment error back, we pass an error
-	// up to main which eventually calls fatal() and returns
-	// with a non-zero exit code.
-	if resp.PaymentError != "" {
-		return errors.New(resp.PaymentError)
-	}
+			// If we get a payment error back, we pass an error up
+			// to main which eventually calls fatal() and returns
+			// with a non-zero exit code.
+			if status.State != routerrpc.PaymentState_SUCCEEDED {
```
Comment (bhandras) [inline cmd/lncli/cmd_pay.go]: It's a bit hard to follow but won't this print the error twice? (once in `printRespJSON` and then wherever we process the error from this call)

File: cmd/lncli/commands.go [hunk 12]
```diff
@@ -2666,6 +2666,33 @@ func queryRoutes(ctx *cli.Context) error {
...
+	case ctx.IsSet("fee_limit"):
+		return &lnrpc.FeeLimit{
+			Limit: &lnrpc.FeeLimit_Fixed{
+				Fixed: ctx.Int64("fee_limit"),
+			},
+		}, nil
+	case ctx.IsSet("fee_limit_percent"):
+		return &lnrpc.FeeLimit{
+			Limit: &lnrpc.FeeLimit_Percent{
+				Percent: ctx.Int64("fee_limit_percent"),
```
Comment (bhandras) [inline cmd/lncli/commands.go]: Shouldn't we do validation here, to make sure that 0 <= percent <= 100?

File: lnrpc/routerrpc/router_server.go (hunk 5 above)
Comment (bhandras) [reply lnrpc/routerrpc/router_server.go]: (i know this is done deal but i still agree with the different PR)

File: cmd/lncli/cmd_pay.go [hunk 13]
```diff
@@ -403,6 +403,53 @@ func sendPaymentRequest(ctx *cli.Context,
 	}
 }
 
+var trackPaymentCommand = cli.Command{
```
Comment (bhandras) [inline cmd/lncli/cmd_pay.go]: Would be nice to have a usage string.

File: cmd/lncli/cmd_pay.go [hunk 14]
```diff
@@ -403,6 +403,53 @@ func sendPaymentRequest(ctx *cli.Context,
...
+	conn := getClientConn(ctx, false)
+	defer conn.Close()
+
+	client := routerrpc.NewRouterClient(conn)
+
+	if !args.Present() {
+		return fmt.Errorf("hash argument missing")
+	}
+
+	hash, err := hex.DecodeString(args.First())
```
Comment (bhandras) [inline cmd/lncli/cmd_pay.go]: Do we want to see if args are provided?

File: cmd/lncli/cmd_pay.go [hunk 15]
```diff
@@ -403,6 +403,53 @@ func sendPaymentRequest(ctx *cli.Context,
 	}
```
Comment (bhandras) [inline cmd/lncli/cmd_pay.go]: nit: as there's multiple calls in this file, maybe better to call it cmd_payments.go ?

File: lnrpc/rpc.proto [hunk 16]
```diff
@@ -503,12 +503,15 @@ service Lightning {
...
-    */
-    rpc SendPayment (stream SendRequest) returns (stream SendResponse);
+    Deprecated, use routerrpc.SendPayment. SendPayment dispatches a
+    bi-directional streaming RPC for sending payments through the Lightning
+    Network. A single RPC invocation creates a persistent bi-directional
+    stream allowing clients to rapidly send payments through the Lightning
+    Network with a single persistent connection.
+    */
+    rpc SendPayment (stream SendRequest) returns (stream SendResponse) {
+        option deprecated = true;
```
Comment (bhandras) [inline lnrpc/rpc.proto]: does this mean the API is still available but users get a notice?

File: cmd/lncli/commands.go (hunk 4 above)
Comment (joostjager) [reply cmd/lncli/commands.go]: yes a reminiscence of k-shortest...

File: cmd/lncli/cmd_pay.go [hunk 17]
```diff
@@ -20,6 +20,12 @@ import (
 	"github.com/urfave/cli"
 )
 
+const (
+	// paymentTimeoutSecs is the default timeout for the payment loop in
+	// lnd. No new attempts will be started after the timeout.
+	paymentTimeoutSecs = 60
```
Comment (bhandras) [inline cmd/lncli/cmd_pay.go]: I'd use `paymentTimeoutSeconds` to conform with other code parts.

File: cmd/lncli/cmd_pay.go (hunk 2 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: I am in doubt about it. It can be done at any time really too. This parameter already existed.

File: cmd/lncli/cmd_pay.go (hunk 3 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: done

File: cmd/lncli/cmd_pay.go (hunk 15 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: My plan was to only put the commands in here that initiate a payment (and track payment, which is an extension of send payment). Therefore I chose the verb "pay". There are more payment-related commands like queryroutes, listpayments, etc. I think it is best to stick to one file per command as much as possible.

File: lnrpc/rpc.proto (hunk 16 above)
Comment (joostjager) [reply lnrpc/rpc.proto]: It is only documentation

File: cmd/lncli/cmd_pay.go [hunk 18]
```diff
@@ -0,0 +1,581 @@
+package main
```
[the remark this answers is not in the document]
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: Agreed

File: cmd/lncli/cmd_pay.go (hunk 10 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: `CltvLimit` is `int32` in the proto

File: cmd/lncli/cmd_pay.go (hunk 11 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: Yes, that is indeed the case. I could also return a more generic error like 'payment failed', to not duplicate the detailed reason. But if we are printing an error line anyway, maybe it is not bad to give the detail as well?

File: cmd/lncli/commands.go (hunk 12 above)
Comment (joostjager) [reply cmd/lncli/commands.go]: Added. I did leave out the upper bound, percentages > 100% could be valid?

File: cmd/lncli/cmd_pay.go (hunk 13 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: Good point, added

File: cmd/lncli/cmd_pay.go (hunk 14 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: At is happening right above, `Args.Present()`

File: cmd/lncli/cmd_pay.go (hunk 15 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: The reason for making an exception to the single command per file rule is that there is code shared.

File: lnrpc/rpc.proto (hunk 16 above)
Comment (joostjager) [reply lnrpc/rpc.proto]: As clarified offline: the API is still available, but users need to prep for its removal.

File: cmd/lncli/cmd_pay.go (hunk 15 above)
Comment (joostjager) [reply cmd/lncli/cmd_pay.go]: Hmm or should I do more splitting?

File: cmd/lncli/cmd_pay.go (hunk 10 above)
Comment (bhandras) [reply cmd/lncli/cmd_pay.go]: Sorry, you're right.

File: cmd/lncli/cmd_pay.go (hunk 14 above)
Comment (bhandras) [reply cmd/lncli/cmd_pay.go]: Indeed, missed that, was looking for the count :)

File: cmd/lncli/cmd_pay.go (hunk 11 above)
Comment (bhandras) [reply cmd/lncli/cmd_pay.go]: I'm OK with this if this is OK, just wanted to take notice here.

File: cmd/lncli/cmd_pay.go (hunk 15 above)
Comment (cfromknecht) [reply cmd/lncli/cmd_pay.go]: my personal preference would be to end up with one file per command, but this is already a good step in that direction so either is fine with me

Comment (bhandras) [reply cmd/lncli/cmd_pay.go]: yeah not a deal breaker, more like a nit. I'm OK with this totally.

File: cmd/lncli/cmd_pay.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (cfromknecht) [reply cmd/lncli/cmd_pay.go]: fwiw a contributor added coverage and unit tests to `lncli` many moons ago, but it was never merged https://github.com/lightningnetwork/lnd/pull/728

[the remark this answers is not in the document]
Comment (bhandras) [reply cmd/lncli/cmd_pay.go]: looks like review never really started for that PR, maybe we could encourage the contributor to resuscitate it. Perhaps some stickers could help :)

**Comment by joostjager** [comment]
Moved itest conversion into separate PR #4132
